File: run-migrations.js
import { Pool } from 'pg';
import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// Supabase PostgreSQL connection detayları
// Supabase Dashboard → Settings → Database → Connection String adresinden alabilirsiniz
const DATABASE_URL = process.argv[2] || process.env.DATABASE_URL;

if (!DATABASE_URL) {
  console.error('❌ DATABASE_URL gerekli!');
  console.error('\nKullanım:');
  console.error('  node run-migrations.js "postgresql://user:password@host:5432/db"');
  console.error('\nVeya:');
  console.error('  export DATABASE_URL="postgresql://..." && node run-migrations.js');
  console.error('\n💡 Connection String nerede bulunur:');
  console.error('  https://app.supabase.com/ → Settings → Database → Connection String');
  process.exit(1);
}

const pool = new Pool({
  connectionString: DATABASE_URL,
  ssl: process.env.NODE_ENV === 'production' 
    ? { rejectUnauthorized: true }
    : { rejectUnauthorized: false }
});

const migrations = [
  '001_initial_schema.sql',
  '002_setup_rls_and_triggers.sql',
  '003_create_auth_trigger.sql',
  '004_create_policies.sql'
];

// SQL statement'ları parse et ve ayrıştır
// Dollar-quoted strings'i dikkate al
function parseSql(sql) {
  const statements = [];
  let current = '';
  let inDollarQuote = false;
  let dollarTag = '';
  let i = 0;

  while (i < sql.length) {
    // Dollar quote check
    if (sql[i] === '$' && !inDollarQuote) {
      let j = i + 1;
      while (j < sql.length && (sql[j].match(/[a-zA-Z0-9_]/) || sql[j] === '$')) {
        j++;
      }
      if (sql[j] === '$') {
        dollarTag = sql.substring(i, j + 1);
        inDollarQuote = true;
        current += dollarTag;
        i = j + 1;
        continue;
      }
    }

    // Check for end of dollar quote
    if (inDollarQuote && sql.substring(i, i + dollarTag.length) === dollarTag) {
      current += dollarTag;
      inDollarQuote = false;
      i += dollarTag.length;
      continue;
    }

    // Semi-colon ends statement (if not in dollar quote)
    if (sql[i] === ';' && !inDollarQuote) {
      current = current.trim();
      if (current && !current.startsWith('--')) {
        statements.push(current);
      }
      current = '';
      i++;
      continue;
    }

    current += sql[i];
    i++;
  }

  // Add remaining statement
  current = current.trim();
  if (current && !current.startsWith('--')) {
    statements.push(current);
  }

  return statements;
}

async function runMigrations() {
  const client = await pool.connect();
  
  try {
    console.log('🚀 Migration işlemi başlanıyor...\n');
    
    for (const migration of migrations) {
      const filePath = path.join(__dirname, 'supabase', 'migrations', migration);
      
      try {
        console.log(`⏳ ${migration} çalıştırılıyor...`);
        
        const sql = fs.readFileSync(filePath, 'utf-8');
        
        // Tüm SQL'i bir kez çalıştır (tüm statements içinde)
        await client.query(sql);
        
        console.log(`✅ ${migration} başarılı\n`);
        
      } catch (err) {
        console.error(`❌ ${migration} hatası:`);
        console.error(`   ${err.message}\n`);
        throw err;
      }
    }
    
    console.log('✨ Tüm migration\'lar başarılı!');
    
  } finally {
    client.release();
    await pool.end();
  }
}

runMigrations().catch(err => {
  console.error('Fatal error:', err.message);
  process.exit(1);
});

File: run_migrations.py
#!/usr/bin/env python3
"""
Supabase migration runner
Supabase veritabanında SQL migration'larını çalıştırır.

Kullanım:
  python run_migrations.py <DB_URL>
  
Örnek:
  python run_migrations.py "postgresql://postgres:password@db.supabase.co:5432/postgres"
  
Connection String nerede bulunur:
  1. https://app.supabase.com/ → Projecte git
  2. Settings → Database → Connection String
  3. 'Connection Details' seçeneğini aç
  4. PostgreSQL connection string'i kopyala
"""

import sys
import os
import glob
import psycopg2
from pathlib import Path

def get_migrations():
    """Migration dosyalarını kronolojik sırada al"""
    migrations_dir = Path(__file__).parent / 'supabase' / 'migrations'
    
    migration_files = [
        '001_initial_schema.sql',
        '002_setup_rls_and_triggers.sql',
        '003_create_auth_trigger.sql',
        '004_create_policies.sql'
    ]
    
    return [migrations_dir / f for f in migration_files if (migrations_dir / f).exists()]

def run_migrations(db_url):
    """SQL migration'larını çalıştır"""
    
    try:
        conn = psycopg2.connect(db_url)
        cursor = conn.cursor()
        
        migrations = get_migrations()
        
        if not migrations:
            print("❌ Migration dosyaları bulunamadı!")
            return False
        
        print(f"🚀 {len(migrations)} migration bulundu.\n")
        
        for migration_file in migrations:
            try:
                with open(migration_file, 'r') as f:
                    sql = f.read()
                
                print(f"⏳ {migration_file.name} çalıştırılıyor...")
                
                cursor.execute(sql)
                conn.commit()
                
                print(f"✅ {migration_file.name} başarılı\n")
                
            except psycopg2.Error as e:
                print(f"❌ {migration_file.name} hatası:")
                print(f"   Error: {e.diag.message_primary if hasattr(e, 'diag') else str(e)}")
                conn.rollback()
                return False
            except Exception as e:
                print(f"❌ {migration_file.name} hatası: {e}")
                conn.rollback()
                return False
        
        cursor.close()
        conn.close()
        
        print("✨ Tüm migration'lar başarılı!")
        return True
        
    except psycopg2.Error as e:
        print(f"❌ Veritabanı bağlantı hatası:")
        print(f"   {e}")
        print("\n💡 Connection string'i kontrol edin:")
        print("   https://app.supabase.com/ → Settings → Database → Connection String")
        return False
    except Exception as e:
        print(f"❌ Hata: {e}")
        return False

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(__doc__)
        print("\n⚠️  Database URL gerekli!")
        print("\nOrnekler:")
        print("  python run_migrations.py 'postgresql://postgres:password@db.supabase.co:5432/postgres'")
        sys.exit(1)
    
    db_url = sys.argv[1]
    
    if run_migrations(db_url):
        sys.exit(0)
    else:
        sys.exit(1)
